pkg/volume/util/fs/fs.go

```diff
@@ -55,63 +59,90 @@ func Info(path string) (int64, int64, int64, int64, int64, int64, error) {
 	return available, capacity, usage, inodes, inodesFree, inodesUsed, nil
 }
 
-// DiskUsage gets disk usage of specified path.
-func DiskUsage(path string) (*resource.Quantity, error) {
+// DiskUsage calculates the number of inodes and disk usage for a given directory
+func DiskUsage(path string) (UsageInfo, error) {
+	var usage UsageInfo
+
+	if path == "" {
+		return usage, fmt.Errorf("invalid directory")
+	}
+
 	// First check whether the quota system knows about this directory
 	// A nil quantity with no error means that the path does not support quotas
 	// and we should use other mechanisms.
-	data, err := fsquota.GetConsumption(path)
-	if data != nil {
-		return data, nil
-	} else if err != nil {
-		return nil, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
-	}
-	// Uses the same niceness level as cadvisor.fs does when running du
-	// Uses -B 1 to always scale to a blocksize of 1 byte
-	out, err := exec.Command("nice", "-n", "19", "du", "-x", "-s", "-B", "1", path).CombinedOutput()
+	consumption, err := fsquota.GetConsumption(path)
 	if err != nil {
-		return nil, fmt.Errorf("failed command 'du' ($ nice -n 19 du -x -s -B 1) on path %s with error %v", path, err)
+		return usage, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
 	}
-	used, err := resource.ParseQuantity(strings.Fields(string(out))[0])
+	inodes, err := fsquota.GetInodes(path)
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse 'du' output %s due to error %v", out, err)
+		return usage, fmt.Errorf("unable to retrieve inode consumption via quota for %s: %v", path, err)
 	}
-	used.Format = resource.BinarySI
-	return &used, nil
-}
 
-// Find uses the equivalent of the command `find <path> -dev -printf '.' | wc -c` to count files and directories.
-// While this is not an exact measure of inodes used, it is a very good approximation.
-func Find(path string) (int64, error) {
-	if path == "" {
-		return 0, fmt.Errorf("invalid directory")
-	}
-	// First check whether the quota system knows about this directory
-	// A nil quantity with no error means that the path does not support quotas
-	// and we should use other mechanisms.
-	inodes, err := fsquota.GetInodes(path)
 	if inodes != nil {
-		return inodes.Value(), nil
-	} else if err != nil {
-		return 0, fmt.Errorf("unable to retrieve inode consumption via quota for %s: %v", path, err)
+		usage.Bytes = consumption.Value()
```

Comment:
should it be if consumption != nil?

Reply:
That is a good question.

Previously the code looked like this:
```go
	// First check whether the quota system knows about this directory
	// A nil quantity with no error means that the path does not support quotas
	// and we should use other mechanisms.
	data, err := fsquota.GetConsumption(path)
	if data != nil {
		return data, nil
	} else if err != nil {
		return nil, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
	}
```
But what happens if `fsquota.GetConsuption(path)` returns a data value that is non-nil _and_ an error? With the above code, it would gladly return the `data` even if there was an error returned. 

Normally in Go code, if there is an error returned, the other value(s) returned are invalid and must be discarded, but I'm not sure that is the case with `fsquota.GetConsumption` and `fsquota.GetInodes`? are there any valid case where those return a value _and_ an error, where we should ignore the error?

My change looks like this (which is difficult to see by just looking at the git diff):
```go
    // First check whether the quota system knows about this directory
    // A nil quantity with no error means that the path does not support quotas
    // and we should use other mechanisms.
    consumption, err := fsquota.GetConsumption(path)
    if err != nil {
        return usage, fmt.Errorf("unable to retrieve disk consumption via quota for %s: %v", path, err)
    }
    inodes, err := fsquota.GetInodes(path)
    if err != nil {
        return usage, fmt.Errorf("unable to retrieve inode consumption via quota for %s: %v", path, err)
    }

    if inodes != nil {
        usage.Bytes = consumption.Value()
    }

    if consumption != nil {
        usage.Inodes = inodes.Value()
    }

    if inodes != nil && consumption != nil {
        return usage, nil
    }
```

So, if either of `fsquota.GetConsumption` or `fsquota.GetInodes` returns error, we abort everything with that error, assuming that the return value can not be used.

Once we have tested that the return values are valid (eg no errors returned) we check if they are nil or not, and if both are non-nil we can return without walking the path.

However, if we are supposed to ignore any errors from either `fsquota.GetConsumption` or `fsquota.GetInodes` as long as the return value is non-nil, then my code is wrong and needs change.

Reply:
>So, if either of fsquota.GetConsumption or fsquota.GetInodes returns error, we abort everything with that error, assuming that the return value can not be used.

In this case, my thought is we should use the filepath.walk approach to get the size instead of returning error directly. I checked quota function, and I don't think there will be case that data!=nil && err!=nil

```go
    // First check whether the quota system knows about this directory
    // A nil quantity with no error means that the path does not support quotas
    // and we should use other mechanisms.
    consumption, err1 := fsquota.GetConsumption(path)
    inodes, err2 := fsquota.GetInodes(path)

    if consumption != nil && inode !=nil {
             usage.Bytes = consumption.Value()
             usaga.Inodes = inode.Value()
            return usage, nil
    }

     ...

    // the following use filepath.Walk to get usage

   ```

Reply:
I think you are right, looks like `fsquota.GetConsumption` returns error in case `xfs_quota` fails to execute, and we dont want return error in that case. I think we simply want fallback to `filepath.walk` on any error.